bridges/snowbridge/primitives/router/src/inbound/mod.rs

```diff
@@ -291,6 +381,106 @@ where
 			[GlobalConsensus(network), AccountKey20 { network: None, key: token.into() }],
 		)
 	}
+
+	/// Constructs an XCM message destined for AssetHub that withdraws assets from the sovereign
+	/// account of the Gateway contract and either deposits those assets into a recipient account or
+	/// forwards the assets to another parachain.
+	fn convert_send_native_token(
+		message_id: H256,
+		chain_id: u64,
+		token_id: TokenId,
+		destination: Destination,
+		amount: u128,
+		asset_hub_fee: u128,
+	) -> Result<(Xcm<()>, Balance), ConvertMessageError> {
+		let network = Ethereum { chain_id };
+		let asset_hub_fee_asset: Asset = (Location::parent(), asset_hub_fee).into();
+
+		let (dest_para_id, beneficiary, dest_para_fee) = match destination {
+			// Final destination is a 32-byte account on AssetHub
+			Destination::AccountId32 { id } =>
+				(None, Location::new(0, [AccountId32 { network: None, id }]), 0),
+			// Final destination is a 32-byte account on a sibling of AssetHub
+			Destination::ForeignAccountId32 { para_id, id, fee } =>
+				(Some(para_id), Location::new(0, [AccountId32 { network: None, id }]), fee),
+			// Final destination is a 20-byte account on a sibling of AssetHub
+			Destination::ForeignAccountId20 { para_id, id, fee } =>
+				(Some(para_id), Location::new(0, [AccountKey20 { network: None, key: id }]), fee),
+		};
+
+		let total_fees = asset_hub_fee.saturating_add(dest_para_fee);
+		let total_fee_asset: Asset = (Location::parent(), total_fees).into();
+
+		let asset_loc =
+			ConvertAssetId::convert(&token_id).ok_or(ConvertMessageError::InvalidToken)?;
+
+		let mut reanchored_asset_loc = asset_loc.clone();
+		reanchored_asset_loc
+			.reanchor(&GlobalAssetHubLocation::get(), &UniversalLocation::get())
+			.map_err(|_| ConvertMessageError::CannotReanchor)?;
+
+		let asset: Asset = (reanchored_asset_loc, amount).into();
+
+		let inbound_queue_pallet_index = InboundQueuePalletInstance::get();
+
+		let mut instructions = vec![
+			ReceiveTeleportedAsset(total_fee_asset.clone().into()),
+			BuyExecution { fees: asset_hub_fee_asset, weight_limit: Unlimited },
+			DescendOrigin(PalletInstance(inbound_queue_pallet_index).into()),
+			UniversalOrigin(GlobalConsensus(network)),
+			WithdrawAsset(asset.clone().into()),
+		];
+
+		let bridge_location = Location::new(2, GlobalConsensus(network));
+
+		match dest_para_id {
+			Some(dest_para_id) => {
+				let dest_para_fee_asset: Asset = (Location::parent(), dest_para_fee).into();
+
+				instructions.extend(vec![
+					// `SetFeesMode` to pay transport fee from bridge sovereign, which depends on
+					//  unspent AH fees deposited to the bridge sovereign,
+					//  more context and analysis in https://github.com/paritytech/polkadot-sdk/pull/5546#discussion_r1744682864
+					SetFeesMode { jit_withdraw: true },
```

Comment:
I also don't like using JIT withdraw here. Who guarantees bridge sovereign on AH has enough funds to cover this?

With XCMv5 transport fees will be handled exclusively from `fees` register and [SetFeesMode will be removed](https://github.com/polkadot-fellows/RFCs/pull/106) so this approach is not future proof.

using half the fee for local exec and half for transport is a _dump solution_ but very simple and should work well in practice

if you want stronger guarantees you need to pass another param coming from ethereum specifying dynamic ratio of split of fees between exec and transport - this way the UI/dApp on Ethereum side can offchain dry-run this message and find accurate costs and provide accurate split

Reply:
> Who guarantees bridge sovereign on AH has enough funds to cover this?

As you point out in the next comment we add `SetAppendix`  to make sure unspent AH fees deposited to the bridge sovereign. 

As analysis in https://github.com/paritytech/polkadot-sdk/pull/5546#discussion_r1744682864 demonstrated the unspent portion should be pretty enough.

Reply:
> For stronger guarantees we can increase the asset_hub_fee to 0.2 DOT(currently 0.1 DOT) which will be charged from user on Etherem.

This works just as well without `SetFeesMode { jit_withdraw: true }`:

You start off with `asset_hub_fee = X + T + B` (eXecution + Transport + Buffer) - with a decent buffer (like you propose for JIT as well) you don't even need to know exact values of X, T or B.

During execution on AH:
1. `X+T+B` DOT is teleported in (holding now X+T+B),
2. `X` is spent on `BuyExecution` (holding now T+B),
3. `T` is spent _from holding_ during execution of `Initiate*` (holding now = B),
4. `B` is deposited to SA during `DepositAll` running in Appendix at the end.

Reply:
https://github.com/paritytech/polkadot-sdk/pull/5546/commits/5203dc02b59d4743bca9633b3b6bcc7e37b7e883